### PR TITLE
fix: sanitize file name when uploading image

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

Sanitizes filename so it does not contain null characters

### CODE REVIEW

**Comment by Marc-Roig** [comment]
this fixes the immediate issue, but the bigger issue is there is an unhandled error in the file streaming that I am not able to identify

**Comment by derrickmehaffy** [comment]
Expirmental: https://github.com/strapi/strapi/actions/runs/8423480135

File: packages/core/upload/server/services/upload.js
Comment (alexandrebodin) [inline packages/core/upload/server/services/upload.js:246]: not sure I understand the reordering 🤔

File: packages/core/upload/server/services/upload.js
Comment (alexandrebodin) [inline packages/core/upload/server/services/upload.js:231]: when generating the thumbnail & responsive sizes are we also running the sanitization ?

**Comment by derrickmehaffy** [comment]
New experimental for changes: https://github.com/strapi/strapi/actions/runs/8426633374

**Comment by derrickmehaffy** [comment]
Exp version: `0.0.0-experimental.e5740babedd53cf5b6af99d74920b6b9ef1e4c11`

**Comment by derrickmehaffy** [comment]
Noticed a new error in the console (though it does fix the immediate issue) we should find a better way to handle this error.

@Marc-Roig I will DM you directly the error I see when fetching the file.

**Comment by Marc-Roig** [comment]
>Noticed a new error in the console (though it does fix the immediate issue) we should find a better way to handle this error.

I believe the error you sent me is unrelated to the upload plugin, and it's caused by koa detecting a malicious path, you get the same error accessing `http://localhost:1337//admin`

We do have a global try catch that should catch unhandled errors, but when errors are tied to streams, they need to be handled in the same stream handlers.

File: packages/core/upload/server/services/upload.js
Comment (Marc-Roig) [reply packages/core/upload/server/services/upload.js:246]: Sanitizing the file name did work for this inmediate issue, but reordering the upload prevented any unhandled errors for some reason, I did not have the time to find which stream was causing the issue. We could take a look together

File: packages/core/upload/server/services/upload.js
Comment (Marc-Roig) [reply packages/core/upload/server/services/upload.js:231]: It only runs once when parsing the original file data, downstream transformations assume the original data is already sanitized.

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy I created another experimental `0.0.0-experimental.69488bcd961d679d6ddfce3b00db926e3e5c713f`
> 
> The current behaviour is preventing the file to be uploaded at all if there are invalid characters.

Sounds good I'll do another test today too confirm everything looks good and ideally we can merge before the branch lock for 4.22.0

**Review by alexandrebodin — APPROVED** [review]

**Comment by derrickmehaffy** [comment]
new exp with latest changes for final validation: https://github.com/strapi/strapi/actions/runs/8439903139

**Comment by Marc-Roig** [comment]
@derrickmehaffy will wait for your approval to merge the pr 🫡

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy will wait for your approval to merge the pr 🫡

Sounds good, exp is here: `0.0.0-experimental.e47108ccbbc4ad1bfaf4526fa6b70d6ace1ca7a9` I'm gonna test it right now, give me about 20 minutes

**Comment by derrickmehaffy** [comment]
@Marc-Roig I was able to locate one more issue, DM'ing you the details with another repo video

**Comment by Marc-Roig** [comment]
@derrickmehaffy thank you for the report! <3 
I fixed that other issue and I am creating another experimental in https://github.com/strapi/strapi/actions/runs/8441431517

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy thank you for the report! <3 
> I fixed that other issue and I am creating another experimental in https://github.com/strapi/strapi/actions/runs/8441431517

Perfect I'll test after I grab some lunch

**Comment by derrickmehaffy** [comment]
@Marc-Roig perfect, works like a charm
image

Exp version for those who want to test: `0.0.0-experimental.88ca4d55bff422eaab16b97b94cb88244832ef39`

**Review by derrickmehaffy — APPROVED** [review]
Passed manual testing :+1:

**Review by alexandrebodin — APPROVED** [review]

**Pull request merged**
